### PR TITLE
GUIReader::registerTypeAndCallBack for std::function and add CustomGUIReader for lua

### PR DESCRIPTION
**Pull request by litefeel**

### CODE REVIEW

**Comment by litefeel** [comment]
[ci rebuild]

**Comment by darkdukey** [comment]
[ci rebuild]

**Comment by litefeel** [comment]
[ci rebuild]

**Comment by litefeel** [comment]
i can build success. why not there?

**Comment by minggo** [comment]
It seems the error exist in lua-binding.

**Comment by litefeel** [comment]
@minggo ?

**Comment by litefeel** [comment]
@minggo i don't know.

**Comment by minggo** [comment]
The error log is http://115.28.134.83:8000/job/cocos-2dx-pull-request-build/4682/node=android_mac/console.
I will ask somebody to check it.

**Comment by minggo** [comment]
I have asked @samuele3hu to take a look.

**Comment by samuele3hu** [comment]
It's a bug of bindings-generator,it didn't consider the value of `r` in 245 line of `tools/bindings-generator/generatory.py`.I suggest that skipping the lua bindings for this functions by setting the 'tools/tolua/cocos2dx_studio.ini'.

**Comment by minggo** [comment]
What did you mean `the value of r`? Will you please modify `tools/tolua/cocos2dx_studio.ini`?

**Comment by samuele3hu** [comment]
En,I will check out this branch and modify it.

**Comment by minggo** [comment]
@lite3 
@samuele3hu fixed lua-binding issue in #7817.

File: cocos/editor-support/cocostudio/CCSGUIReader.h [hunk 1]
```diff
@@ -85,8 +90,8 @@ class CC_STUDIO_DLL GUIReader : public cocos2d::Ref
     ParseObjectMap _mapObject;
     
 public:
-    ParseCallBackMap getParseCallBackMap() { return _mapParseSelector; };
-    ParseObjectMap getParseObjectMap() { return _mapObject; };
+    ParseCallBackMap* getParseCallBackMap() { return &_mapParseSelector; };
```
Comment (zilongshanren) [inline cocos/editor-support/cocostudio/CCSGUIReader.h:93]: hi @lite3 
Why should we change std::map from value to reference?
I think c++11 std::map has move semantic , it won't introduce extra copy performance penalty here.

File: cocos/base/ObjectFactory.cpp [hunk 2]
```diff
@@ -92,8 +106,13 @@ Ref* ObjectFactory::createObject(const std::string &name)
 	do 
 	{
 		const TInfo t = _typeMap[name];
-		CC_BREAK_IF(t._fun == nullptr);
-		o = t._fun();
+        if (t._fun != nullptr)
```
Comment (zilongshanren) [inline cocos/base/ObjectFactory.cpp:109]: @lite3 
Please fix the indentation, thanks.

File: cocos/editor-support/cocostudio/CCSGUIReader.h (hunk 1 above)
Comment (litefeel) [inline cocos/editor-support/cocostudio/CCSGUIReader.h:93]: @andyque 

```
auto map = reader->getParseCallBackMap();
```
1.  this map is copy, reader->_mapParseSelector. can not modify.
2.  move semantic used by temporary object. but the temporary map need malloc.

**Comment by litefeel** [comment]
@minggo the same problem at build failed. 
fix it at https://github.com/cocos2d/bindings-generator/pull/131

**Comment by samuele3hu** [comment]
@lite3 , I check out your branch,and use the latest bindings-generator for cocos2dx_studio.ini, the result is successful.Why did you modify the bindings-generator.

**Comment by litefeel** [comment]
@samuele3hu 

for bindings-generator, I write it at https://github.com/cocos2d/bindings-generator/pull/131

now, i write it again.

old regex:  `function<(.+) .*\((.*)\)>`
now regex: `function<(.+) *\((.*)\)>`

input string is `function<Ref *(void)>`
result about first group:
old:  `Ref`
now:  `Ref *`

**Comment by litefeel** [comment]
@minggo please review

**Comment by minggo** [comment]
Ok, i have asked @samuele3hu to review lua-binding codes.
Thanks.

**Comment by samuele3hu** [comment]
@lite3,you are right.If use `function<(.+) .*\((.*)\)>`,the parse error about `ret_type` was triggered,through it don't have effect now,thanks.
@minggo , for lua-binding codes,i think it is ok.

**Pull request merged**